File: rs2-core/src/services/sms.rs
//! `sms` service: the canonical HTTP surface over the [`SmsGateway`] capability
//! (PRD §9.2). The provider is swappable per mount via `store.adapter`
//! (`code:<name>@<version>` or `infra:<name>`); the service is provider-agnostic
//! and runs unchanged over Twilio, AWS SNS, … — only the adapter differs.
//!
//! - `POST /<mount>/send` with `{ "to": "+1…", "body": "…" }` → `201 { "id": … }`
//! - `GET  /<mount>/status/<id>` → `200` provider-shaped delivery status.

use async_trait::async_trait;
use http::{Method, StatusCode};
use serde_json::json;

use super::{Service, ServiceContext};
use crate::error::RsError;
use crate::message::Message;

#[derive(Default)]
pub struct SmsService;

impl SmsService {
    pub fn new() -> Self {
        SmsService
    }
}

#[async_trait]
impl Service for SmsService {
    async fn handle(&self, mut msg: Message, ctx: &ServiceContext) -> Result<Message, RsError> {
        let sms = ctx.sms.as_ref().ok_or_else(|| RsError::capability_denied("sms"))?;

        let method = msg.method.clone();
        let segs: Vec<String> = msg.url.service_segments().iter().map(|s| s.to_string()).collect();
        let parts: Vec<&str> = segs.iter().map(String::as_str).collect();

        match (&method, parts.as_slice()) {
            (&Method::POST, ["send"]) => {
                let body = msg
                    .body
                    .as_mut()
                    .ok_or_else(|| RsError::bad_request("POST /send requires a JSON body {to, body}"))?;
                let payload = body.as_json(ctx.limits.materialized_body_bytes).await?;
                let to = payload
                    .get("to")
                    .and_then(|v| v.as_str())
                    .ok_or_else(|| RsError::bad_request("'to' (string) is required"))?;
                let text = payload
                    .get("body")
                    .and_then(|v| v.as_str())
                    .ok_or_else(|| RsError::bad_request("'body' (string) is required"))?;
                let id = sms.send(to, text).await?;
                Ok(msg.response(StatusCode::CREATED, Some(crate::message::Body::from_json(&json!({ "id": id })))))
            }
            (&Method::GET, ["status", id]) => {
                let status = sms.status(id).await?;
                Ok(msg.ok_json(&status))
            }
            _ => Err(RsError::bad_request(
                "sms endpoint: POST /send {to, body}, GET /status/{id}",
            )),
        }
    }
}

File: rs2-worker/src/services/sms.ts
// `sms` service: the canonical HTTP surface over the SMS gateway capability
// (PRD §9.2). Port of `rs2-core/src/services/sms.rs`. The provider is
// swappable per mount via `store.adapter`; on this host every adapter form
// is refused at build until P4b (cloudflare.md §C.1), so the routes below
// only run once resident adapters arrive — they are ported now so the
// service surface matches Rust byte for byte.
//
// - `POST /<mount>/send` with `{ "to": "+1…", "body": "…" }` → `201 { "id": … }`
// - `GET  /<mount>/status/<id>` → `200` provider-shaped delivery status.

import { Body } from "../runtime/body";
import { RsError } from "../runtime/error";
import type { Message } from "../runtime/message";
import type { Service, ServiceContext } from "./context";

export class SmsService implements Service {
  async handle(msg: Message, ctx: ServiceContext): Promise<Message> {
    const sms = ctx.sms;
    if (!sms) throw RsError.capabilityDenied("sms");

    const parts = msg.url.serviceSegments();

    if (msg.method === "POST" && parts.length === 1 && parts[0] === "send") {
      if (!msg.body) throw RsError.badRequest("POST /send requires a JSON body {to, body}");
      const payload = await msg.body.asJson(ctx.limits.materializedBodyBytes);
      const obj = payload && typeof payload === "object" && !Array.isArray(payload) ? payload : {};
      const to = obj.to;
      if (typeof to !== "string") throw RsError.badRequest("'to' (string) is required");
      const text = obj.body;
      if (typeof text !== "string") throw RsError.badRequest("'body' (string) is required");
      const id = await sms.send(to, text);
      return msg.response(201, Body.fromJson({ id }));
    }
    if (msg.method === "GET" && parts.length === 2 && parts[0] === "status") {
      const status = await sms.status(parts[1]!);
      return msg.okJson(status);
    }
    throw RsError.badRequest("sms endpoint: POST /send {to, body}, GET /status/{id}");
  }
}
